File: Python3/601 - 800/664 Strange Printer.py
# 396ms, 70.67%; 13.4MB, 33.33%
class Solution:
    def __init__(self):
        self.mat = [[0] * 100 for _ in range(100)]

    def strangePrinter(self, s: str) -> int:
        ss = []
        for c in s:
            if not ss or ss[-1] != c: ss.append(c)
        return self.dp(ss, 0, len(ss) - 1)

    def dp(self, s: str, l: int, r: int) -> int:
        if l > r or self.mat[l][r]: return self.mat[l][r]
        self.mat[l][r] = 1 + self.dp(s, l, r - 1)
        for i in range(l, r):
            if s[i] == s[r]:
                self.mat[l][r] = min(self.mat[l][r], self.dp(s, l, i) + self.dp(s, i + 1, r - 1))
        return self.mat[l][r]

File: 601 - 800/664. Strange Printer.cc
// 16ms, 100.0%; 9.3MB, 100.0%
#include <bits/stdc++.h>
using namespace std;

static int x = []() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    return 0;
}();

class Solution {
private:
    int f[100][100];

public:
    int strangePrinter(string &s) {
        memset(f, 0, sizeof(f));
        s.erase(unique(s.begin(), s.end()), s.end());
        return dp(s, 0, s.size() - 1);
    }

private:
    int dp(const string &s, int l, int r) {
        if (l > r) return 0;
        if (f[l][r]) return f[l][r];
        f[l][r] = 1 + dp(s, l, r - 1);
        for (int i = l; i < r; ++i)
            if (s[i] == s[r]) f[l][r] = min(f[l][r], dp(s, l, i) + dp(s, i + 1, r - 1));
        return f[l][r];
    }
};
